### DISCUSSION TITLE
Wonderful tool!

### DISCUSSION THREAD

**Discussion by philuxem** [SilverGreen93/CDPExplorer#8] [Ideas]
Hi,

First of all, thank you very much for this very useful tool that I've extensively and successfully used over the past couple of months, processing tens of thousands of files.

I thought it could be useful to share some comments with you:

1) I noticed a minor (non-blocking) "functional deficiency". Although it is possible to add *.cdp2 files via drag & drop, it is not an available option when using the "Add CDP" menu.

2) Not sure that it would useful for many but the ability to retrieve assets from old *.cmpa backup files could possibly also be considered. I would consider this a nice-to-have feature only.

3) The feature that I would really appreciate to obtain is the ability to export individual asset files with an incremental suffix added to the asset name when there are name collisions. Indeed, the sole options of either overwriting an existing asset or skipping an export do not allow me to ensure that I do not miss an error-free asset. I already came across cases where I exported 30+ different (i.e., not exact duplicate) occurrences of the same asset. The only way to achieve this with the current version is to process CDP files one by one, each in a distinct folder and this is dissuasive.

Hope it makes sense to you.
Happy to discuss more the topic if you wish.

Best regards,
philuxem

**Comment by SilverGreen93** [comment]
Hi philuxem,
Thank you for using this software and making time to share your thoughts with me.

1. I don't have any cdp2 files to test it and confirm it works correctly. I know cdp2 files were only used by Trainz Classics releases, which I do not own. Please provide a simple sample cdp2 files containing 1-2 assets if possible to test myself. If testing confirms they work, will include the file filter.
2. I considered cmpa support multiple times, but the format is not as straight forward to reverse engineer as the cdp was. I don't consider it a priority because anyone can open a cmpa with Trainz 12 or older and export the files to cdp from there, and the effort to implement it will be too big. CMPA files are less and less common as they are not in use since the release of TANE.
3. This I will do. Seems a reasonable ask and might find it useful for myself as well. Will create a separate issue for this to be tracked.

Thanks,
Mihai

**Reply by philuxem** [reply]
Hi Mihai,

Thanks for your answer.

1. I successfully processed 900+ .cdp2 files via drag & drop. As I understand from your message that cdp explorer was not intentionally designed to process .cdp2 files, I changed the extension of some of those files from .cdp2 to .cdp and I could process them as well. So, I wonder now whether there is any difference between .cdp and .cdp2 file formats. Please find below a sample list: 

http://trainz-railroad.ucoz.ru/_ld/0/26_rustramjunction.cdp2
http://trainz-railroad.ucoz.ru/_ld/0/28_nasepgravel.cdp2
http://trainz-railroad.ucoz.ru/_ld/0/29_RUSTRAMULKS.cdp2
http://trainz-railroad.ucoz.ru/_ld/0/30_RUSTRAMtrackNZ.cdp2
http://trainz-railroad.ucoz.ru/_ld/0/35_NovoistrinskV2..cdp2
http://trainz-railroad.ucoz.ru/_ld/0/38_Backgroundspack.cdp2
http://trainz-railroad.ucoz.ru/_ld/0/41_volchanskayaKS.cdp2
http://trainz-railroad.ucoz.ru/_ld/0/43_Rustram_opora_k.cdp2
http://trainz-railroad.ucoz.ru/_ld/0/50_71-40171-402_BE.cdp2
http://trainz-railroad.ucoz.ru/_ld/0/53_6cF.cdp2
http://trainz-railroad.ucoz.ru/_ld/0/54_Pavlovsk_beta.cdp2
http://trainz-railroad.ucoz.ru/_ld/0/55_Samara_map_ses.cdp2
http://trainz-railroad.ucoz.ru/_ld/0/60_Hruschevskiepya.cdp2
http://trainz-railroad.ucoz.ru/_ld/0/61_Tablichka_3D_bu.cdp2
http://trainz-railroad.ucoz.ru/_ld/0/66_6eG.cdp2
http://trainz-railroad.ucoz.ru/_ld/0/67_KBSH.cdp2
http://trainz-railroad.ucoz.ru/_ld/0/70_pack.cdp2
http://trainz-railroad.ucoz.ru/_ld/0/71_TR_KurortnajaMo.cdp2
http://trainz-railroad.ucoz.ru/_ld/0/72_T3x.cdp2
http://trainz-railroad.ucoz.ru/_ld/0/76_cCb.cdp2
http://trainz-railroad.ucoz.ru/_ld/0/89_1_ozm.cdp2
http://trainz-railroad.ucoz.ru/_ld/0/94___50_.cdp2
http://trainz-railroad.ucoz.ru/_ld/0/96_16_.cdp2
http://trainz-railroad.ucoz.ru/_ld/0/97_4Qv.cdp2
http://trainz-railroad.ucoz.ru/_ld/0/98_Rflora_Heracliu.cdp2
http://trainz-railroad.ucoz.ru/_ld/1/102_tatra_pack.cdp2
http://trainz-railroad.ucoz.ru/_ld/1/120_TR_road_pack1.cdp2
http://trainz-railroad.ucoz.ru/_ld/1/124_S_Novokrasinsk_.cdp2
http://trainz-railroad.ucoz.ru/_ld/1/136__65_.cdp2
http://trainz-railroad.ucoz.ru/_ld/1/146_N_3.cdp2
http://trainz-railroad.ucoz.ru/_ld/1/179____.cdp2
http://trainz-railroad.ucoz.ru/_ld/2/218_AE86_KVR_TM1_PA.cdp2
http://trs2004-2006.ucoz.net/_fr/0/11.cdp2
http://trs2004-2006.ucoz.net/_fr/0/12.cdp2
http://trs2004-2006.ucoz.net/_fr/0/13.cdp2
http://trs2004-2006.ucoz.net/_fr/2/24.cdp2
http://trs2004-2006.ucoz.net/_ld/0/27_alex_pack_2.cdp2
http://trs2004-2006.ucoz.net/_ld/0/29_alex_pack_3.cdp2
http://trs2004-2006.ucoz.net/_ld/0/30_september09_sta.cdp2
http://trainz.cz/download/okoli/silnice-a-mosty/675-dalnice-viceproude
http://trainz.cz/download/okoli/silnice-a-mosty/1229-mosty-viceproude#
http://roubiaro.fr/app/download/11679010028/maison+construction+brique.cdp2
http://roubiaro.fr/app/download/11680043728/PACK+ROQUEFORT.cdp2

2. I fully concur with your analysis. I indeed use TS12 to import content of .cmpa files when needed. If the format is too complex, then it is not worth the effort.

3. That would be great, thank you! 

BTW, how could I contribute to your work effort?

Best regards,
Philippe

**Reply by SilverGreen93** [reply]
Thanks for the links. Will test myself the cdp2 files and if all works, will add support in next release.

For CMPA support, if anyone knows or can reverse engineer the file format specification and share me the details, I would be happy to implement it.

Thanks for you interest in supporting this project. Best way to contribute would be help coding for adding more features, but I can also accept donations via credit card here: https://revolut.me/mihai4oh (in any supported currency).

**Comment by philuxem** [comment]
I'm not a developer but could participate in testing if it helps.
